exp/topology/desiredstate/desired_state_test.go

```diff
@@ -3545,25 +3572,50 @@ func TestCalculateRefDesiredAPIVersion(t *testing.T) {
 
-func validateCleanupCluster(req runtimehooksv1.RequestObject) error {
-	var cluster clusterv1beta1.Cluster
-	switch req := req.(type) {
-	case *runtimehooksv1.BeforeClusterUpgradeRequest:
-		cluster = req.Cluster
-	case *runtimehooksv1.AfterControlPlaneUpgradeRequest:
-		cluster = req.Cluster
-	default:
-		return fmt.Errorf("unhandled request type %T", req)
-	}
+func validateClusterParameter(originalCluster *clusterv1.Cluster) func(req runtimehooksv1.RequestObject) error {
+	// return a func that allows to check if expected transformations are applied to the Cluster parameter which is
+	// included in the payload for lifecycle hooks calls.
+	return func(req runtimehooksv1.RequestObject) error {
+		var cluster clusterv1beta1.Cluster
+		switch req := req.(type) {
+		case *runtimehooksv1.BeforeClusterUpgradeRequest:
+			cluster = req.Cluster
+		case *runtimehooksv1.AfterControlPlaneUpgradeRequest:
+			cluster = req.Cluster
+		default:
+			return fmt.Errorf("unhandled request type %T", req)
+		}
 
-	if cluster.GetManagedFields() != nil {
-		return errors.New("managedFields should have been cleaned up")
-	}
-	if _, ok := cluster.Annotations[corev1.LastAppliedConfigAnnotation]; ok {
-		return errors.New("last-applied-configuration annotation should have been cleaned up")
-	}
-	if _, ok := cluster.Annotations[conversion.DataAnnotation]; ok {
-		return errors.New("conversion annotation should have been cleaned up")
+		// check if managed fields and well know annotations have been removed from the Cluster parameter included in the payload lifecycle hooks calls.
+		if cluster.GetManagedFields() != nil {
+			return errors.New("managedFields should have been cleaned up")
+		}
+		if _, ok := cluster.Annotations[corev1.LastAppliedConfigAnnotation]; ok {
+			return errors.New("last-applied-configuration annotation should have been cleaned up")
+		}
+		if _, ok := cluster.Annotations[conversion.DataAnnotation]; ok {
+			return errors.New("conversion annotation should have been cleaned up")
+		}
+
+		// check the Cluster parameter included in the payload lifecycle hooks calls has been properly converted from v1beta2 to v1beta1.
+		// Note: to perform this check we convert the parameter back to v1beta2 and compare with the original cluster +/- expected transformations.
+		v1beta2Cluster := &clusterv1.Cluster{}
+		if err := cluster.ConvertTo(v1beta2Cluster); err != nil {
+			return err
+		}
+
+		originalClusterCopy := originalCluster.DeepCopy()
+		originalClusterCopy.SetManagedFields(nil)
+		if originalClusterCopy.Annotations != nil {
+			annotations := maps.Clone(cluster.Annotations)
+			delete(annotations, corev1.LastAppliedConfigAnnotation)
+			delete(annotations, conversion.DataAnnotation)
+			originalClusterCopy.Annotations = annotations
+		}
+
+		if !apiequality.Semantic.DeepEqual(originalClusterCopy, v1beta2Cluster) {
+			return errors.Errorf("call to extension is not passing the expected cluster object: %s", cmp.Diff(originalClusterCopy, v1beta2Cluster))
```

Comment:
```suggestion
			return fmt.Errorf("call to extension is not passing the expected cluster object: %s", cmp.Diff(originalClusterCopy, v1beta2Cluster))
```

Reply:
We have >400 usages of errors.Errorf and >200 of fmt.Errorf and an open issue about what we're going to do about  pkg/errors.

At this point I would not try to force usage of one or the other in unit tests